test/jdk/java/util/zip/GZIP/GZIPInputStreamConcat.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8322256
+ * @summary Test support for concatenated gzip streams
+ * @run junit GZIPInputStreamConcat
+ */
+
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.io.*;
+import java.util.*;
+import java.util.stream.*;
+import java.util.zip.*;
+
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+
+public class GZIPInputStreamConcat {
+
+    private int bufsize;
+
+    public static Stream<Object[]> testScenarios() throws IOException {
+
+        // Test concat vs. non-concat, garbage vs. no-garbage, and various buffer sizes on random data
+        Random random = new Random();
+        final ArrayList<List<Object>> scenarios = new ArrayList<>();
+        for (int size = 1; size < 1024; size += random.nextInt(32) + 1) {
+            scenarios.add(List.of(randomData(0, 100), size));
+        }
+        return scenarios.stream().map(List::toArray);
+    }
+
+    @ParameterizedTest
+    @MethodSource("testScenarios")
+    public void testScenario(byte[] uncompressed, int size) throws IOException {
+        this.bufsize = size;
+        testScenario(uncompressed);
+    }
+
+    public void testScenario(byte[] uncompressed) throws IOException {
+
+        // Compress the test data
+        byte[] compressed = deflate(uncompressed);
+
+        // Decompress a single stream with no extra garbage - should always work
+        byte[] input = compressed;
+        byte[] output = uncompressed;
```

Comment:
The assignment of `input` and `output` variables in this method add a lot of noise. Could this be simplified to something like this?

```java
        // Compress the test data
        byte[] compressed = deflate(uncompressed);

        // Decompress a single stream with no extra garbage - should always work
        testDecomp(compressed, uncompressed);

        // Decompress a truncated GZIP header
        testDecompFail(oneByteShort(gzipHeader()), EOFException.class);

        // Decompress a single stream that is one byte short - should always fail
        testDecompFail(oneByteShort(compressed), EOFException.class);

        // Decompress a single stream with one byte of extra garbage (trying all 256 possible values)
        for (int extra = 0; extra < 0x100; extra++) {
            testDecomp(oneByteLong(compressed, extra), uncompressed);
        }

        // Decompress a single stream followed by a truncated GZIP header
        testDecomp(concat(compressed, oneByteShort(gzipHeader())), uncompressed);

        // Decompress a single stream followed by another stream that is one byte short
        testDecompFail(concat(compressed, oneByteShort(compressed)), IOException.class);

        // Decompress two streams concatenated
        testDecomp(concat(compressed, compressed), concat(uncompressed, uncompressed));

        // Decompress three streams concatenated
        testDecomp(concat(compressed, compressed, compressed),
                concat(uncompressed, uncompressed, uncompressed));

        // Decompress three streams concatenated followed by a truncated GZIP header
        testDecomp(concat(compressed, compressed, compressed, oneByteShort(gzipHeader())),
                concat(uncompressed, uncompressed, uncompressed));
```